AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStoreAdmin/src/Mutations.dfy

```diff
@@ -0,0 +1,662 @@
+  // Mutations Token is valid
+  // logicalKeyStoreName is valid
+  function ValidateApplyMutationInput(
+    input: Types.ApplyMutationInput,
+    logicalKeyStoreName: string
+  ): (output: Result<Types.ApplyMutationInput, Types.Error>)
+    ensures output.Success? ==> |logicalKeyStoreName| > 0 && ValidateMutationToken(input.mutationToken).Success?
+  {
+    var _ :- ValidateMutationToken(input.mutationToken);
+    :- Need(|logicalKeyStoreName| > 0,
+            Types.KeyStoreAdminException(message := "LogicalKeyStoreName cannot be empty!"));
+    Success(input)
+  }
+
+  // Ensures:
+  // Branch Key ID is set
+  // Original is set (not going to deseriazlize yet)
+  // Terminal is Valid (not going to deseriazlize yet)
+  function ValidateMutationToken(
+    input: Types.MutationToken
+  ): (output: Result<Types.MutationToken, Types.Error>)
+    ensures output.Success?
+            ==>
+              && |input.Identifier| > 0
+              && |input.Original| > 0
+              && |input.Terminal| > 0
+  {
+    :- Need(|input.Identifier| > 0,
+            Types.KeyStoreAdminException(message := "Branch Key Identifier cannot be empty!"));
+    :- Need(|input.Original| > 0,
+            Types.KeyStoreAdminException(message := "Original cannot be empty!"));
+    :- Need(|input.Terminal| > 0,
+            Types.KeyStoreAdminException(message := "Terminal cannot be empty!"));
+    Success(input)
+  }
+
+  method {:vcs_split_on_every_assert} ApplyMutation(
+    input: Types.ApplyMutationInput,
+    logicalKeyStoreName: string,
+    keyManagerStrategy: keyManagerStrat,
+    storage: Types.AwsCryptographyKeyStoreTypes.IKeyStorageInterface
+  )
+    returns (output: Result<Types.ApplyMutationOutput, Types.Error>)
+    requires ValidateApplyMutationInput(input, logicalKeyStoreName).Success?
+    requires
+      && storage.ValidState()
+      && keyManagerStrategy.ValidState()
+    ensures
+      && storage.ValidState()
+      && keyManagerStrategy.ValidState()
+    modifies
+      storage.Modifies,
+             match keyManagerStrategy
+             case reEncrypt(km) => km.kmsClient.Modifies
+             case decryptEncrypt(kmD, kmE) => kmD.kmsClient.Modifies + kmE.kmsClient.Modifies
+    requires keyManagerStrategy.reEncrypt?
+  {
+
+    var keyManager := keyManagerStrategy.reEncrypt;
+    // assume {:axiom} storage.Modifies !! keyManager.kmsClient.Modifies;
+
+    // -= Query for page Size Branch Key Items
+    var queryOut? := storage.QueryForVersions(
+      Types.AwsCryptographyKeyStoreTypes.QueryForVersionsInput(
+        exclusiveStartKey := input.mutationToken.ExclusiveStartKey,
+        Identifier := input.mutationToken.Identifier,
+        // DDB TransactWrite takes a max of 25
+        // with the mutation lock, 24 items is the maximum we could write
+        // TODO this size should be optional
+        // DDB is best when you write slow and steady, not as fast as possible :)
+        pageSize := input.pageSize.UnwrapOr(24)));
```

Comment:
@seebees once suggested the pageSize should be 3,
as that is what the Key Store uses for Create.
This is not a Beta blocker, but we should resolve this
```suggestion
        pageSize := input.pageSize.UnwrapOr(24)));
```

Reply:
Lowering Default to 3.
Will add doc comments and Example.